config_center/nacos/client.go

```diff
@@ -0,0 +1,218 @@
+package nacos
+
+import (
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+import (
+	"github.com/nacos-group/nacos-sdk-go/clients"
+	"github.com/nacos-group/nacos-sdk-go/clients/config_client"
+	nacosconst "github.com/nacos-group/nacos-sdk-go/common/constant"
+	perrors "github.com/pkg/errors"
+)
+
+import (
+	"github.com/apache/dubbo-go/common/constant"
+	"github.com/apache/dubbo-go/common/logger"
+)
+
+const logDir = "logs/nacos/log"
+
+// NacosClient Nacos client
+type NacosClient struct {
+	name       string
+	NacosAddrs []string
+	sync.Mutex // for Client
+	client     *config_client.IConfigClient
+	exit       chan struct{}
+	Timeout    time.Duration
+	once       sync.Once
+	onceClose  func()
+}
+
+// Client Get Client
+func (n *NacosClient) Client() *config_client.IConfigClient {
+	return n.client
+}
+
+// SetClient Set client
+func (n *NacosClient) SetClient(client *config_client.IConfigClient) {
+	n.Lock()
+	n.client = client
+	n.Unlock()
+}
+
+type option func(*options)
+
+type options struct {
+	nacosName string
+	client    *NacosClient
+}
+
+// WithNacosName Set nacos name
+func WithNacosName(name string) option {
+	return func(opt *options) {
+		opt.nacosName = name
+	}
+}
+
+// ValidateNacosClient Validate nacos client , if null then create it
+func ValidateNacosClient(container nacosClientFacade, opts ...option) error {
```

Comment:
Is possible that this method is invoked concurrently? If so, if-null and then create will cause some problem.

Reply:
may be invoked concurrently, so it has a lock in ``` container.SetNacosClient(newClient) ``` and ``` container.NacosClient().SetClient(&client) ```.  Finally added if-null logic .